### PR TITLE
Columns autosize

### PR DESCRIPTION
**Pull request by SatanEnglish**
Created an auto column sizer for resizing when you cannot preset widths for the data.

Provided an example of how it works.

### CODE REVIEW

**Comment by SatanEnglish** [comment]
@6pac @ghiscoding 
I've noticed this one has other commits to it that shouldn't be in here.
Can you guys  just have a look at how it works and give some feed back then I'll fix the pull to be just the **autocolumsize plugin** commit like it should be.

**Comment by ghiscoding** [comment]
@SatanEnglish 
You might have a name conflict since there's already a function named `autosizeColumns` in the core lib, as can be seen [here](https://github.com/6pac/SlickGrid/blob/master/slick.grid.js#L2045)

Also that would be helpful to post maybe an animated gif of what this feature does? What does this plugin do that the function `autosizeColumns` doesn't do?

**Comment by SatanEnglish** [comment]
There is an example added for this

**Comment by SatanEnglish** [comment]
Shouldn't be a name conflict as it's a plugin.
It's got the same naming as it does the same sort of thing.
Anyway check out the example it's.

example-autocolumnsize.html in this branch

**Comment by SatanEnglish** [comment]
Also here is a quick explanation.
Say you have all dynamic columns and data.

This plugin takes into account both the data and the column and finds the max width needed for the column.

e.g. Quick gif to try and help

**Comment by ghiscoding** [comment]
So this is an Auto Size by Content plugin. Thanks for the animated gif, I'm more visual and this helps a lot to understand what this plugin can do. I haven't totally looked at your code, but are you doing an entire `for` loop to analyze the possible width to use or just what is in the buffer? If you have 50k rows (or worst a million row, which SlickGrid can handle), that could have an impact. If that is the case, you should tell the user that it's doing that, maybe add a note in your example (top right) about this impact. 

Taking a quick look at your code, it looks like you are doing an actual loop through the entire data set (from DataView), that will have a huge impact on big data set. However, you're actually doing double the work, you're re-copying everything in a temp array on this [line](https://github.com/6pac/SlickGrid/pull/323/files#diff-55102cae390a730ebc2862c97eda877dR135), meaning that if your data set is 1 million rows, you'll end up with 2 million rows in the cache..not good. It's preferable to use a temp `maxWidth` variable and analyze it on the spot. Basically you set the `maxWidth` to 0 at first and whenever the current row is wider than that `maxWidth`, you replace it (so the first row will always replace the 0). 

Considering that you're analyzing the entire data set, isn't it better to just check what is in the cache? Or at least limit the number of rows to analyze, you could say check only first 1000 rows and that wouldn't be too bad. Then if you scroll to a row that is wider, you could recall the function to calculate but starting from that row number and using previous `maxWidth` as the new check (say you find the next wider column at row 2010, you could recalculate from row 2010 to 3010).

**Comment by SatanEnglish** [comment]
> I haven't totally looked at your code, but are you doing an entire for loop to analyze the possible width to use or just what is in the buffer
There are 2 cases for the sizing.

1. The field text is checked for the biggest of all the data for a column and we find the biggest and then create a fake cell with all the formatting of a normal cell and the width is taken.
2. It's if it's not text it loads the formatter for the row into a fake cell.
3. It ignore columns that are not resizable

Currently that's the best optimization I could manage It does have a max size to stop at to which helps with speed a little to.

"Noted though will put a warning in final version"


> you're re-copying everything in a temp array
It's getting just the data for each column as it needs to run that against it's formatter to find it's actual size.
e.g. when looking at duration column it's an array of just duration values.
Might be able to make it that better though haven't really looked into it enough around that part.
Again i've sped this up as much as possible but for formatters that adjust the data excessively it won't completely work
e.g. value = 1 and the formatter generates `<span>This is what 1 means</span>` but that can be in a later feature to force computed value through formatter.

This was orginally based off of [link](https://github.com/naresh-n/slickgrid-column-data-autosize)
But I found it to slow and adapted it to get as much speed improvements as I could.

It also calculates widths once and stores to sizes so that calling it with when more columns are added we don't need to recalc() when resizeCanvas is called and then adjusting screen size.
When it's asked for I can add a force resize if need be.


> you could say check only first 1000 rows
I've got 20k of rows in my table with 20 columns and this runs in milliseconds on all.
While it may pay to do something on massive sets of data it's a lot faster than you would expect.

**Comment by SatanEnglish** [comment]
Also thanks for the feedback you have a least given me somewhere I can try and make an improvement

**Comment by 6pac** [comment]
My 2c: I have considered this before and I think you need an option for the analysis type. eg: 
   AutosizeColumnsOnContentMode: None (0), FirstRow(-1), AllLoadedRows (-2), FirstNRows (n) 
This could possibly be per-column, ie. a column option. You might also want to provide an alternative formatter just for sizing that works faster than the regular formatter and returns a size rather than a string.

**Comment by 6pac** [comment]
Also creating and destroying the temp cells is going to be expensive.  i'm pretty sure you could find a library or function that can provide the relative text size for a string, and then scale that by just creating a single cell with sample text and determining the ratio of abstracted 'text size' to displayed size. Again, perhaps an option to determine the algorithm?

**Comment by SatanEnglish** [comment]
The text cell version of this takes the string with the longest length and then adds a buffer.
That ones really quite quick it's cheating a bit but improves the speed a lot.

I've had no speed issues.

I'll get a version with the large amount of rows running so we can bench mark some tests.

There is a double click function on the resize that's not working with your guys version. (Not sure whats changed)

Also there is a resize by Name.
This allows for single columns to be resized.

AutosizeColumnsOnContentMode is a nice idea. (My case doesn't need it)

**Comment by SatanEnglish** [comment]
Ok more than 8k seems to have a big speed issue that isn't on my version.
I'll see what over speed improvements will help.
N rows and recalculate sounds like it would be nice.

**Comment by 6pac** [comment]
This feature has been on my mind for a while. I will have time to pay it and related features some attention in about 3 months, I think (I'd like to make sure that all the options for autosizing are integrated somewhat elegantly).

As such, I'd be willing to pull this as a first iteration, but I might give you a bit more time to refine it as you see fit. It's great that it's a plugin, that make it much more attractive to pull.

**Comment by SatanEnglish** [comment]
I put the code back to my orginal version for now.
Maybe we can put a note on it once I've tweeked it that it's a ALPHA version.
For now that only works for Default columns e.g. Text columns.

It's not to slow with this revert back to not calculating no text cells.

@6pac I'll also make it into another branch so as to not have other commits in it

**Comment by 6pac** [comment]
Great. If you do those things, I'll see if I can look at it and do some tweaking of my own over the next week or so.

**Comment by SatanEnglish** [comment]
I found something to speed it up even more however it uses a unique value approach.
Amazing for text cases but won't be good for some formatters.

            let length = texts.length, result = [], seen = new Set();
            outer:
            for (let index = 0; index < length; index++) {
                let value = texts[index];
                if (seen.has(value)) continue outer;
                seen.add(value);
                result.push(value);
            }

very fast way to get unique data. [Link](https://medium.com/@jakubsynowiec/unique-array-values-in-javascript-7c932682766c ) around the 36ms mark for 200k of rows.  

Increased my speed a fair bit as I have columns like sex with MALE, FEMALE in every column.

Anyway I'll put this in and redo the example I have my end with the 500k rows to see if i'm better off.

**Comment by SatanEnglish** [comment]
On the 500K example:

Without unique check 64664.30000006221 milliseconds.
With unique 20736.29999998957 milliseconds.

**Comment by SatanEnglish** [comment]
Next moved
Line 202 `context.font = element.css("font-size") + " " + element.css("font-family");`
To be set once I think I know what this is doing so moving it should be fine.

Before 20736.29999998957 milliseconds.
After 6533.000000054017 milliseconds.

**Comment by SatanEnglish** [comment]
I'm missing something to get the double click to work any idea what it is?
Plugin OR Javascript?

**Comment by SatanEnglish** [comment]
After more randomized data
6789.0999999362975 milliseconds.

**Comment by SatanEnglish** [comment]
@6pac
 I feel that the fiddling around was worth it here.
Nearly 10x faster than I had it this morning.

Leaving it for now but the basic version here is starting to look promising to me.
Working with other formatter that don't work with the field the column is attached to will be the real fun here.

**Comment by SatanEnglish** [comment]
Added way to put rows into url example to test more quantities.

Test 5 time with each as random data
1K -> 24ms, 18ms, 18ms, 18ms, 18ms,
10K -> 189ms, 181ms, 179ms, 192ms, 190ms
100K -> 1316ms,1322ms,1312ms,1315ms,1325ms
1M rows -> 13760ms, 13772ms, 14244ms, 13557ms, 13679ms,

@6pac here are some more times for you.

**Comment by 6pac** [comment]
Really, the 100K is borderline and the 1M is unacceptable. However I think generally the first 1000 rows will be indicative enough for the column width. It's true that for a lot of lookup columns, date columns, numeric columns etc the range or values will be quite low. I still think that a little more smarts around the per-column  algorithm setting would work wonders.

**Comment by SatanEnglish** [comment]
In my case the 100K - 200K time was acceptable as this is only called once and the user already has to wait for a loading gif for a while anyway.

1M is unacceptable yep is quite slow there.
Having a some nice was to pass quicker lookups per column would definitely help.

From having a play we can at least see there is a lot of gain to have.

Anyway I cannot spend any more time on this for a while so unless I find a change while working on other stuff I'll leave it for you to have a play with next when you have time.

PS: I had to change a line to get this working due the the frozen columns.
`var gridCanvas = $(_grid.getContainerNode()).find(".grid-canvas")[0];`
So I think I'm only checking the left side of the 2 canvases.
(This will need looking into but didn't exist and time of creation)

**Comment by SatanEnglish** [comment]
@6pac  Have you managed to have a play with this yet?

**Comment by 6pac** [comment]
Sorry no :-) Was on extended holidays, now very busy again catching up on the work front!  
However I will promise to have a look this weekend, this is a worthy addition.

**Comment by 6pac** [comment]
OK am checking this out offline. Will take another week probs.

**Comment by 6pac** [comment]
Here's my proposed API. See what you think. It's a bit ugly in how it relates to the underlying column, I've kind of ignored that and assumed that if the plugin is installed, it will override whatever is there. Perhaps need some more thought to integrate it more cleanly.

    // AutoColumnSize Overview
    // -----------------------
    // The AutoColumnSize plugin is designed to allow the grid to make intelligent choices about the 
    // width of its columns based on the cell content of the header and rows. The columns widths can  
    // be sized independently of the viewport, or a fit-viewport-to-columns or fit-columns-to-viewport  
    // strategy can be used.
    // The column width of each column can be locked to a particular width, set to a specific guide
    // width, or based on the content of the header (optionally) and any number of rows. Specific  
    // columns may be set to expand to fill extra space, or all columns may be scaled to fit the viewport.
    //
    // AutoColumnSize plugin options
    // -----------------------------
    //  option
    //   .viewportMode: FitColsToViewport (NoFit, FitColsToViewport, FitViewportToCols)
    //   .minViewportWidthPx: undefined
    //   .maxViewportWidthPx: undefined
    //
    //  DETAILS
    //
    //  viewportMode
    //    - a named mode (details below) defining the relationship between the viewport and the total 
    //      column width
    //
    //  minViewportWidthPx, maxViewportWidthPx
    //    - the minimum viewport width in the case of viewportMode.FitViewportToCols
    //
    //
    // ViewportModes
    // -------------
    // 
    // NoFit:
    //   - columns are sized independently of the viewport width. There will be empty space at the
    //     right of the viewport if the columns are smaller, and a scroll bar if they are larger.
    //   - SizeToRemaining is ignored, its presence will trigger a console message
    //
    // FitColsToViewport:
    //   - columns sized are calculated using the column strategies
    //   - if addl space remains in the viewport and there are SizeToRemaining cols, just the  
    //     SizeToRemaining cols expand proportionally to fill viewport
    //   - otherwise (ie. no SizeToRemaining cols or viewport smaller than columns) all cols other 
    //     than 'Locked' scale in proportion to fill viewport
    //
    // FitViewportToCols:
    //   - columns sized are calculated using the column strategies
    //   - viewport is resized to fit columns
    //   - if viewport with is outside MinViewportWidthPx and MaxViewportWidthPx, then the viewport is
    //     set to MinViewportWidthPx or MaxViewportWidthPx and the FitColsToViewport algorithm is used
    //
    //
    // AutoColumnSize column options (declared as members of an 'autosize' property)
    //     
    //   .ignoreHeaderText: false
    //   .colValueArray: undefined
    //   .formatterOverride: undefined
    //   .widthPx: undefined
    //   .minWidthPx: undefined
    //   .maxWidthPx: undefined
    //   .autoWidthStrategy: TopNRows (Locked, Guide, Top1Row, TopNRows, AllRows)
    //   .sizeToRemaining: undefined
    //   .checkRowCount: 100
    //
    //  DETAILS
    //
    //  ignoreHeaderText
    //    - ignores the content of the header when looking for the widest content
    //
    //  colValueArray
    //    - an array of all possible values for the column (for drops downs, etc). Used instead of row
    //      data. May also be used to pass a single value (eg. longest form of date)
    //
    //  formatterOverride
    //    - allows the formatter to be overridden with a more efficient function for examining the width.
    //      The highest of the following will be used: (1) data value converted to text, (2) formatter, 
    //      (3) formatter override
    //
    //  widthPx
    //    - the column width used in the case of AutoWidthStrategy.Locked or AutoWidthStrategy.Guide
    //
    //  minWidthPx, maxWidthPx
    //    - the absolute maximum and minimum column widths for all modes
    //
    //  autoWidthStrategy
    //    - a named strategy (see below) for determining the column width
    //
    //  sizeToRemaining
    //    - designate a column to expand to fill space in viewportMode.FitColsToViewport. If multiple 
    //      columns are marked with sizeToRemaining, columns will fill unused space in proportion to 
    //      their initial calculated size
    //
    //  checkRowCount
    //    - in AutoWidthStrategy.TopNRows, the number of rows to check for widest content.
    //
    //
    // AutoWidthStrategy
    // -----------------
    //
    // Locked:
    //   - the final column width is locked to exactly .widthPx and will not be scaled.
    //     MinWidthPx and MaxWidthPx should not be used and will trigger a log message if specified.
    //
    // Guide:
    //   - the .widthPx will be used as an initial width, but the column may be scaled
    //
    // Top1Row:
    //   - content width of the first row only will be used as the column width. This is useful where
    //     all column entries share a standard format, eg dates, GUIDs
    //
    // TopNRows:
    //   - content width of the widest of the first .checkRowCount rows will be used as the column width.
    //
    // AllRows:
    //   - content width of the widest of all the rows will be used as the column width.
    //

**Comment by SatanEnglish** [comment]
@6pac 
It's a bit hard to get my head around this.
Any chance you could give image examples even if it's just paint as to what combinations do?

As far as I can tell from reading the modals I think you have covered most of the cases I can think of.

Some small questions/suggestions

// ViewportModes
NoFit: <= Do we need another like this that will Fit to the view port if to small but will still resize normally if going to take up more space? Even if it just treats the last column like its  "sizeToRemaining" type.

// AutoWidthStrategy
// Top1Row: <== The first Row with information in it?? 
e.g Lets say I have dates but it can be blank I want the first row with data to be what the size is taken off.

What settings would you see for doing approximately what it's doing now. E.g. using it's own distinct text method and sizing off the distinct returned list?

PS: I'll get back to you with how to integrate if I come up with something.

**Comment by 6pac** [comment]
> // ViewportModes
> NoFit: <= Do we need another like this that will Fit to the view port if to small but will still resize normally if going to take up more space? Even if it just treats the last column like its "sizeToRemaining" type.

Not sure what you mean by this. I think your plugin only offers NoFit at the moment? There are potentially some more cases around resorting to a scrollbar in FitColsToViewport and FitViewportToCols if the columns are much too wide.  
It comes down to the use case. The idea was to have defaults that allow it to do well automatically with anything that's thrown at it, but also allow options to fine tune performance and behaviour.

> // AutoWidthStrategy
> // Top1Row: <== The first Row with information in it??
> e.g Lets say I have dates but it can be blank I want the first row with data to be what the size is taken off.

Good call.

> What settings would you see for doing approximately what it's doing now. E.g. using it's own distinct text method and sizing off the distinct returned list?

At the moment I think your version uses NoFit and the AllRows strategy.

> PS: I'll get back to you with how to integrate if I come up with something.

No expecting you to do this work, I'm happy to put all these options together, you've done great work optimising the actual width measurements.  
The reason I propose all these options is that in my mind, it's always more complex than just sizing to content. Often you want a fit to viewport function. There are some (long text columns) for example, where you definitely don't want to size to content, you want a standard or a specifically designed guide size. This kind of column is a great candidate for expanding to fit a viewport. It's about getting the most flexibility and greatest speed for least configuration.

**Comment by SatanEnglish** [comment]
@6pac done anything on this lately? (No rush just interested on updates)

**Comment by 6pac** [comment]
Making steady progress. Most of my time is going on creating a really useful demo page. Hopefully should have something ready by start of next week.

**Comment by ghiscoding** [comment]
@6pac 
I just realized you asked for my comments... There are mainly 2 things that sticks out for me.

1. I still would prefer to have a different name for the plugin to avoid confusion with the `autosizeColumns` method that already exist in the core lib, I think the plugin could be named `AutoContentColumnSize` or something similar.
2. You have a `formatterOverride` in your documentation plugin, however in recent PR #346 we have a similar override that we decided to name `canRowBeSelected` while my previous choice of that same method was `selectableOverride`. I'm wondering if I should go back to that name or if we should have some kind of name convention for these override methods? 
   - Note that since we didn't release a new version yet, we still have time to rename the override method in the merged PR #346 and the upcoming #351 which is similar. I actually kinda like to see the override word somewhere in the name, should we rename...?

**Comment by 6pac** [comment]
1. good call
2. I do think the 'override' convention is better where something replaces an existing function/property with essentially the same functionality, so I'd agree with the rename.

**Comment by 6pac** [comment]
Oooookay .... this is complicated. I didn't realise how much the sizing is wrapped up with the core grid code. So I'd advise you (@SatanEnglish) to go ahead with the plugin independent of me for now, if it's useful for you. 

I'll work through this and probably integrate the different sizing strategies into the core grid code. But it's not going to be quick.  
As I have probably hinted, I find the current autosizing behaviour quite cumbersome, so I think this is worth doing really well and thoroughly.

**Comment by SatanEnglish** [comment]
@6pac
No rush I've already got this in my own version as a separate plugin so It's no issue for me to leave mine as is for now.

I am however interested WHY you think it's wrapped up tightly with core code tho.

From what I've played with it it seems to be quite self contained.
My thoughts on how to implement this would of been a follows.

1) Assume the developer using it can put extra fields on the columns something along the lines of.
```
Columns[0].SizeInfo{
AutoWidthStrategy: Locked,
ViewportModes: NoFit
}
```
2) The plugin could have defaults or ignore Items that don't have the SizeInfo
3) Supply a method that generates defaults onto each column.
4) Have an options for min, max width where SizeInfo not present.
5) Have a force resize function so we can store if the column size has been calculated.

1 <-- This allows us to determine everything from the column information.


Anyway maybe there is something I'm not thinking of that you have noticed that couples it with CORE.

Interested in your feed back

**Comment by 6pac** [comment]
it's just that ```autosizeColumns()```, which is where all the main decision making is done about the column arrangement, is in ```slick.grid.js```.  ... and the column max/min/width are all part of that logic, which is out of the plugin's control. ... and the current code assumes the grid has been rendered, where it would be much better to be able to sort this out pre the first render.  ... AND there is manual resizing and all the logic and safeguards around that.

All up, the plugin as it stands works well - if just using the existing column sizing strategy. I was hoping for rather a lot more than that!   
Not that any of this is impossible, it's just going to be a bit tricky.

**Comment by SatanEnglish** [comment]
@6pac  I see good being able to do it instead of standard part would be good and save processing.
Wonder if there is a way to write the plugin so that we could pull it into the core once we have time.

Anyway I can see what you mean about it being a big job if your planning that

**Comment by 6pac** [comment]
I have made a lot of progress, but busy with work at the moment, not a free hour. Sometime in the next few weeks I'll get to it and release something.

**Comment by ninezero90hy** [comment]
@6pac It's an amazing plugin. I really need it. 😄

File: plugins/slick.autocolumnsize.js [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+        function resizeAllColumns() {
+            if (_grid.getContainerNode().offsetParent === null) return;
+            var allColumns = _grid.getColumns();
+            var gridOps = _grid.getOptions();
+
+            allColumns.forEach(function (columnDef, index) {
+                if (columnDef && (!columnDef.resizable || columnDef._autoCalcWidth === true)) return;
+                var el = document.getElementById(_grid.getUID() + columnDef.id);
+                var headerWidth = getElementWidth(el);
+                var autoSizeWidth = Math.max(headerWidth, getMaxColumnTextWidth(columnDef, index)) + 1;
```
Comment (ninezero90hy) [inline plugins/slick.autocolumnsize.js]: @SatanEnglish @6pac  
Changes to the 'autoSizeWidth' value occur twice unnecessarily.
I think we can improve the code.

- Before
```
var autoSizeWidth = Math.max(headerWidth, getMaxColumnTextWidth(columnDef, index)) + 1;

// Check we are smaller than the maxWidth if provided
autoSizeWidth = Math.min(maxWidth, autoSizeWidth);
```

- After
```
// Check we are smaller than the maxWidth if provided
var autoSizeWidth = Math.min(
	maxWidth, 
	Math.max(headerWidth, getMaxColumnTextWidth(columnDef, index)) + 1
);
```

**Comment by 6pac** [comment]
@ninezero90hy  just FYI, I'm writing something that is a lot larger than the PR here, and so much of this PR probably won't be used (I've definitely used it as a departure point, but hacked it up pretty throughly!). None of this is to detract from @SatanEnglish 's work - it's great but I had plans for something with an even larger scope. Perhaps I'm overthinking it, but I agree that it's very useful - I spend a lot of my dev time tweaking column sizes, and frankly I never want to touch one again, so I want to nail it. He may be willing to share the plugin with you now - I assume that's what is in this PR, but he might have something a little more fancy up his sleeve :-) I'm just not able to guarantee that that plugin will still be useful after my changes. I'll preserve backwards compatibility with existing features, but there will be a lot of overlap with the plugin.

**Comment by ninezero90hy** [comment]
> @ninezero90hy just FYI, I'm writing something that is a lot larger than the PR here, and so much of this PR probably won't be used (I've definitely used it as a departure point, but hacked it up pretty throughly!). None of this is to detract from @SatanEnglish 's work - it's great but I had plans for something with an even larger scope. Perhaps I'm overthinking it, but I agree that it's very useful - I spend a lot of my dev time tweaking column sizes, and frankly I never want to touch one again, so I want to nail it. He may be willing to share the plugin with you now - I assume that's what is in this PR, but he might have something a little more fancy up his sleeve :-) I'm just not able to guarantee that that plugin will still be useful after my changes. I'll preserve backwards compatibility with existing features, but there will be a lot of overlap with the plugin.

Yes, I see. :) 😄

**Comment by SatanEnglish** [comment]
@6pac Good to hear your onto this let me know if you end up needing any help.
@ninezero90hy Glad to see interest in this sorry it's not done yet but as @6pac says it would be great to get it right first time and working well.

**Comment by SatanEnglish** [comment]
@6pac Is this pull now redundent as you have everything you need from the code?
If so I'll close this.

Also how you getting on with the re-write?

**Comment by TimBailey-pnk** [comment]
@6pac I have autosizing requriement where I would like to have some fixed width columns, some percentage and then the others "fill evenly"....
Hopefully once your core changes are done I can just write a plugin?

**Comment by 6pac** [comment]
I am in the final stages of testing the above Autosize functionality. Will hopefully release something in the next few days.

**Comment by 6pac** [comment]
OK this is in the current repo now. I haven't done a release yet, will wait for a few days and sort out a few other minor issues too. 
The Autosizing is no longer a plugin - it's part of the core.

For more details, see [here](https://github.com/6pac/SlickGrid/wiki/Auto-Column-Sizing) in the wiki. Note there's a comprehensive example as well.

**Comment by SatanEnglish** [comment]
@6pac Awsome will have a read in the morning

**Pull request closed without merging**